flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -1358,4 +1358,19 @@ public void actionsUnavailable() throws InterruptedException {
 			mailbox.putMail(actionUnavailableLetter);
 		}
 	}
+
+	private class TimerInvocationContext implements SystemProcessingTimeService.ScheduledCallbackExecutionContext {
```

Comment:
I don't think that the mailbox is exposed to the `SystemProcessingTimeService` at all here. On the contrary.

The `TimerInvocationContext` is in fact a "mailbox sender" (implementation detail, that doesn't matter for the timer service). I'm not using the real `MailboxSender` interface, because it would involve somehow adapting `ProcessingTimeCallback` interface to `Runnable`. Or changing all implementers of `ProcessingTimeCallback` to new api - this may be quite involving and may conflict with others' changes.

Imo, things like `SystemProcessingTimeService` can be neutral about use cases. The fact that it already had checkpoint lock and error handler callback, I find as incidental, rather by design.
Having additional transformation of `ProcessingTimeCallback` in the time service, into a mailbox letter (`Runnable`), would make things even more coupled.

An alternative could be, that there are no newly introduced `SystemProcessingTimeService.ScheduledCallbackExecutionContext`, but keeping `SystemProcessingTimeService` implementation clean (not coupled with `StreamTask`).
Add a proxy `ProcessingTimeService` implementation that would be exposed for others to register timers, but upon registration, the proxy would register it's own callback, which when fired, places the original `ProcessingTimeCallback` as a letter into the mailbox.

Reply:
To clarify. With this change, the `ProcessingTimeCallback` becomes non relevant to `SystemProcessingTimeService`, it can be substituted by any class (e.g. just `Object`) - this object is just passed on timer registration as data and returned back by the time service when the timer fires.

The real callback interface now becomes `ScheduledCallbackExecutionContext`.

Reply:
I think I cannot quiet follow your argument here. My problem with this solution from a design point is, that there is code in the stream task that is talking about how the timer service is wants to enqueue things into the mailbox. From a high-level view, do you agree that this code should not belong there? Depending on what you think about this general concern, there is also a solution that does all this inside the timer service, and the timer service just knows the sender interface of the queue (and maybe the error handler). You could almost move the code into the timer service, one the sender interface is a member there. Does that solve the problem of why you did not want to do this?

Reply:
> My problem with this solution from a design point is, that there is code in the stream task that is talking about how the timer service wants to enqueue things into the mailbox.

It's not "how the timer service wants". But it's rather that the `StreamTask`'s use case, how to react to timer events. Optionally, this helper class can be moved outside of `StreamTask`, but then it have to have reference to checkpoint lock, `AsyncExceptionHandler` and `MailboxSender` from the `StreamTask`.

> From a high-level view, do you agree that this code should not belong there? 

Imo, such code should not belong to timer service, but is fine to be in `StreamTask`, as it's more coupled with the latter (`StreamTask`).

>  Depending on what you think about this general concern, there is also a solution that does all this inside the timer service, and the timer service just knows the sender interface of the queue (and maybe the error handler). You could almost move the code into the timer service, one the sender interface is a member there.

Yes, this is basically to revert the change and additionally add `MailboxSender` into the timer service.
It's similar to moving `TimerInvocationContext` into standalone class, and then inlining it in the timer service, as the only implementation.

> Does that solve the problem of why you did not want to do this?

Personally, I'm against of what you propose. As I've mentioned before, it would make (wrong) things more coupled.

Reply:
I think I cannot follow your arguments. For me it looks obviously wrong to read code ala `invoke(ProcessingTimeCallback callback, long timestamp)` and `callback.onProcessingTime(timestamp)` inside the context of stream task. If we cannot come to an agreement on this, let's ask a third person for another opinion @pnowojski ?